libdnf/rpm/package_query.cpp

```diff
@@ -225,25 +225,69 @@ static inline bool name_arch_compare_lower_solvable(const Solvable * first, cons
 }  //  namespace
 
 
-PackageQuery::PackageQuery(const BaseWeakPtr & base, Flags flags, bool empty) : PackageSet(base), flags(flags) {
-    if (empty) {
-        return;
+PackageQuery::PackageQuery(const BaseWeakPtr & base, Flags flags, bool empty)
+    : PackageSet(base),
+      flags(flags),
+      p_pq_impl(new PQImpl) {
+    if (!empty) {
+        *p_impl |= base->get_rpm_package_sack()->p_impl->get_solvables();
     }
 
+    auto & pool = get_pool(base);
     switch (flags) {
-        case Flags::IGNORE_EXCLUDES:
-            // TODO(jmracek) add exclude application
         case Flags::APPLY_EXCLUDES:
-        case Flags::IGNORE_MODULAR_EXCLUDES:
+            base->get_rpm_package_sack()->p_impl->recompute_considered_in_pool();
+            if (!empty && pool.is_considered_map_active()) {
+                *p_impl &= pool.get_considered_map();
+            }
+            break;
         case Flags::IGNORE_REGULAR_EXCLUDES:
-            *p_impl |= base->get_rpm_package_sack()->p_impl->get_solvables();
+        case Flags::IGNORE_MODULAR_EXCLUDES: {
+            auto considered = base->get_rpm_package_sack()->p_impl->compute_considered_map(flags);
+            if (considered) {
+                p_pq_impl->considered_cache = std::move(considered);
+                if (!empty) {
+                    *p_impl &= *p_pq_impl->considered_cache;
+                }
+            }
             break;
+        }
+        case Flags::IGNORE_EXCLUDES:;
     }
 }
 
 PackageQuery::PackageQuery(libdnf::Base & base, Flags flags, bool empty)
     : PackageQuery(base.get_weak_ptr(), flags, empty) {}
 
+PackageQuery::PackageQuery(const PackageQuery & src)
+    : PackageSet(src),
+      flags(src.flags),
+      p_pq_impl(new PQImpl(*src.p_pq_impl)) {}
+
+PackageQuery::PackageQuery(PackageQuery && src) noexcept
+    : PackageSet(src),
+      flags(src.flags),
+      p_pq_impl(new PQImpl(std::move(*src.p_pq_impl))) {}
```

Comment:
Why do you need to define the move constructor / assignment operator? Can't use the generated ones?

Reply:
Our code moves `*p_pq_impl`.
However, the automatically generated code moves `p_pq_impl`. It's faster. But in `src` there will be` p_pq_impl == nullptr`. Not a problem now, but ...

Reply:
But you are moving it, I don't understand your argument. We use generated move constructors for all our pimpls AFAIK.

And actually your version seems even more dangerous, because `src` after this will contain a "valid" `unique_ptr` which will contain an object that has been moved and can be in an undefined state.

These boilerplate constructor and assignment operators are outright bad and best avoided, they just add potential for mistakes. It's bad enough we need to define the copy ones (which could be avoided if we had a `unique-ptr`-like type that would be copyable by calling the contained type copy constructor).

Reply:
"We use generated move constructors for all our pimpls AFAIK." No, look e.g. at PackageSet.
Why? It uses the same logic as STL. Moved out (src) object may be undefined but valid.
E.g.
`auto dst == std::move(src_packageset);`
Because `src_packageset` is vallid we can call `src_packageset.size()` and get zero.
We can also reuse some moved out packages.
But whit the default generated move operator it will fail. Because it calls `p_impl->size()`. And `p_impl` will be `nullptr`.